### ISSUE TITLE
WaitingList implementation

### ISSUE THREAD

**Issue by jlaumon** [RichieSams/FiberTaskingLib#3]
Hi

I also made an implementation of this fiber tasking system at work (but unfortunately, it's not opensource). Anyway, thanks for your work, it helped me to avoid a lot of mistakes!

I ended up with a different waiting list implementation which I think requires shorter locking (or no locking at all depending on how it's done). I'll try to explain it here, in case it helps.

In your implem, in the `FiberStart` loop, every worker thread checks the counter of every waiting task. But in theory, you only need to check the tasks that wait on a counter when you decrement this particular counter.

To do that, you need to store a list a waiting tasks per counter instead of having a global one. Let's say the counter becomes a struct that contains both its value, and a list a of waiting tasks, you can, every time you decrement it, check the list of waiting tasks and move the one that need to be woken up to a "wake up" list. Then at the beginning of the `FiberStart` loop, you only need to try to dequeue from this list.

**Comment by RichieSams** [comment]
Ah! Very smart. I like it!

I'll try to implement that this afternoon / weekend.

**Comment by RichieSams** [comment]
Deferred until Boost::Context fibers are stable. ie: issue #8

**Comment by RichieSams** [comment]
While this is a good idea in concept, I don't know how well it would fit in the current architecture. I'm going to close the issue for now. Perhaps in the future, it can be re-examined and re-opened.
